woocommerce/woocommerce
# Product Name Suggestions with AI

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

This PR adds the API and the UI for generating improved product titles using AI. I've added a `CompletionService` class here that currently connects to OpenAI, but we can easily change to the Jetpack API once we're ready.

It currently checks that the title is at least 10 characters or the description is at least 50 characters before sending a request to the API. This is to ensure the results are relevant.

I've tried to optimize the prompt to get good enough results, but there is still room for improvement. Currently, the prompt is a bit long, but I've found that it's better to describe in detail what we want so that GPT doesn't go off-track. 

Getting a JSON response was a challenge, and there might still be times (rarely) when the returned response is not valid JSON. I've added a simple retry mechanism where it will ask the API again for a JSON response. But it is nowhere near perfect. For now, it works, and based on my trial runs, I usually get a JSON response %99 of the time.

I've added a JS function to scrape the product data from the edit product page of WooCommerce. Since WooCommece does not provide that data natively to React in the legacy editor, I had to go through the HTML inputs and tags to get that data. This is reusable for any other features we might add to that page.

We send the product's current title, description, categories, tags, and other attributes (set in the attributes tab) to the API, and the suggestion will take all of these into account. The more data is provided, the better the results.

For loading messages, I've added several random messages just to pass the time since the function takes a long time to complete (usually more than 8s).

I'm sure there's tons of room for improvement on the UI/UX, but I think this is a good start.

Closes #38390 
Closes #38391

### Screenshots

Initial state:
<img width="1234" alt="image" src="https://github.com/woocommerce/woocommerce/assets/73110514/8eb91571-d567-4e18-857e-98c7bf38c56a">


Loading screen:
<img width="1221" alt="image" src="https://github.com/woocommerce/woocommerce/assets/73110514/aa86a89a-6372-451c-8fcc-d3ce8228d157">

Suggestions list:
<img width="1219" alt="image" src="https://github.com/woocommerce/woocommerce/assets/73110514/4ff91103-4430-491b-ae1e-6137c54406b5">




### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout branch and link woo-ai plugin to your local.
2. Activate the woo-ai plugin.
3. Ensure your local has the OPEN_AI_KEY defined.
4. Navigate to Products -> Add New
5. Click on the Get Suggestions button and confirm that you are shown an error.
6. Write some description for the product (more than 50 chars) and try again
7. Confirm the suggestions are relevant to the description
8. Add categories, tags, and other attributes and try again to get better suggestions
9. Click suggest alternatives button and confirm alternative suggestions are displayed
